plugins/coreboot/fu-coreboot-sysfs.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2019 9elements Agency GmbH <patrick.rudolph@9elements.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <stdio.h>
+
+#include "fu-plugin-coreboot.h"
+
+#define SYSFS_BUS "/sys/bus/coreboot/"
+
+/* Tries to detect the 'coreboot' kernel module presence. */
+gboolean
+fu_plugin_coreboot_sysfs_probe (void)
+{
+	return g_file_test(SYSFS_BUS, G_FILE_TEST_EXISTS|
+				      G_FILE_TEST_IS_DIR);
+}
+
+/* iterates over sysfs directories until the given ID is found. */
+static gchar *
+fu_plugin_coreboot_find_sysfs (const guint id,
```

Comment:
Is it not possible to rely upon the kernel creating a uevent and udev picking it up instead to find this?

Reply:
The coreboot kernel module creates virtual devices for every coreboot table entry. Looking at the udev code in fwupd I'm not sure how to set up an event listener on a specific bus. As there's no device file, vendor or product id and the subsystem is none of pci, mmc, usb, hid, ....
Can you give an example how such a code would look like?

Reply:
Maybe. Can you show us a tree of what the devices look like. I don't know if the coreboot tree is going to be suitable for the FuUdevDevice model where each exposed device is backed just by one GUdevDevice.

Reply:
We could certainly support a new subsystem for coreboot if we need to too

Reply:
The virtual devices have attributes that needs to be read:

> /sys/bus/coreboot/devices/.../attributes/data
> /sys/bus/coreboot/devices/.../attributes/id

The cbmem driver (if loaded) creates an additional folder on those virtual devices:

> /sys/bus/coreboot/devices/.../cbmem_attributes/data
> /sys/bus/coreboot/devices/.../cbmem_attributes/id
> /sys/bus/coreboot/devices/.../cbmem_attributes/address

The key is matched what is present in *id*, and then contents of *data* is loaded.

Besides the newly added attributes the virtual device folder is empty.

Reply:
I would expect you should be able to pick uevents for those devices and then use the udev path to point to the `data`, `id`, and `address` directories.  I think it should be cleaner than trying to walk the filesystem manually.

Reply:
Do you have a code example how to pick up the uevents for those devices? I've never worked with uevents before.

Reply:
If there's a valid udev subsystem you can do https://github.com/fwupd/fwupd/blob/master/plugins/optionrom/fu-plugin-optionrom.c#L18 and then something like `fu_plugin_set_device_gtype (plugin, FU_TYPE_OPTIONROM_DEVICE);` -- this makes fwupd create an object of the correct type whenever that udev subsystem is added. https://github.com/fwupd/fwupd/blob/master/plugins/optionrom/fu-optionrom-device.c gives a good example of using `fu_udev_device_get_sysfs_path()` for example.

Reply:
I don't think uevents work here.